crates/ra_assists/src/assists/early_return.rs

```diff
@@ -0,0 +1,185 @@
+use hir::db::HirDatabase;
+use itertools::Itertools;
+use ra_syntax::{ast::Block, ast::FnDef, ast::IfExpr, ast::LoopExpr, ast::ReturnExpr, AstNode};
+
+use crate::{Assist, AssistCtx, AssistId};
+
+pub(crate) fn convert_to_guarded_return(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let if_expr: IfExpr = ctx.node_at_offset()?;
+    let expr = if_expr.condition()?.expr()?;
+    let then_block = if_expr.then_branch()?.block()?;
+    if if_expr.else_branch().is_some() {
+        return None;
+    }
+
+    let parent_block = if_expr.syntax().parent()?.ancestors().find_map(Block::cast)?;
+
+    if parent_block.syntax().last_child()? != *if_expr.syntax() {
```

Comment:
I think it's better to use typed API here: `parent_block.expr()? != if_expr.clone().into()`. We probably should add some convenience `PartialEq` impls to avoid explicit upcasting via into...